docker/tarfile/src.go

```diff
@@ -35,14 +38,63 @@ type layerInfo struct {
 	size int64
 }
 
-// NewSource returns a tarfile.Source for the specified path.
-func NewSource(path string) *Source {
+// NewSource returns a tarfile.Source for the specified path or inputStream.
+// Cannot set path and inputStream at the same time.
+// inputStream is mainly used for the docker-daemon transport.
+func NewSource(path string, inputStream io.ReadCloser) (*Source, error) {
 	// TODO: We could add support for multiple images in a single archive, so
 	//       that people could use docker-archive:opensuse.tar:opensuse:leap as
 	//       the source of an image.
+
+	if path != "" && inputStream != nil {
+		return nil, errors.Errorf("Cannot set both path and inputStream")
+	}
+
+	// Check if path is a gzip file and read its contents if it is
+	isGzip := false
+	if path != "" {
+		file, err := os.Open(path)
+		if err != nil {
+			return nil, err
+		}
+		defer file.Close()
+
+		reader, err := gzip.NewReader(file)
+		if err != nil {
+			logrus.Debugf("%q not gzipped", path)
+		} else {
+			defer reader.Close()
+			isGzip = true
+			inputStream = reader
+		}
+	}
+
+	if isGzip || inputStream != nil {
+		// FIXME: use SystemContext here.
+		// Save inputStream to a temporary file
+		tarCopyFile, err := ioutil.TempFile(tmpdir.TemporaryDirectoryForBigFiles(), "docker-tar")
```

Comment:
(Nothing removes `tarCopyFile` in this package – see the other comment in `docker/daemon`.)

Reply:
Moved and being deleted in Source.Close() in docker/tarfile